### PR TITLE
Additional functionality and improvements in the NormData class. 

### PR DESCRIPTION
**Pull request by eminSerin**
- Added functions to save and load the NormData class in NetCDF format (i.e., .nc file). This method is significantly more efficient and faster than saving as a .csv file, particularly for large datasets.
- Implemented the TODO for the "register_batch_effects" function, which now utilizes vectorized operations to improve performance.

### CODE REVIEW

**Comment by AuguB** [comment]
Thanks, I understand. However, the attributes that are reconstructed are not necessarily identical to the ones that were removed. This is because the data in those dicts doesn't necessarily correspond with the data in the arrays, and intentionally so. For instance, when you load data into a NormData object and do a train-test split, both those splits receive identical copies of the dicts, which is intentional and necessary for several functions including proper data synthesis and plotting. When the data is then saved and loaded as .nc, the dicts will be re-estimated on the data in the arrays, and will not match the dicts that were removed. This is easy to test for yourself. A workaround could be to store all those dicts in a pickle next to the .nc file, and load them back into the object after creation. To ensure .pickle and .nc match the same NormData, we could create filenames with a uuid. 

About the other comment, I agree that the toolbox could be improved for large-scale datasets, and this is a straightforward way to do it. All I'm thinking is whether it isn't easier to save a subset of the NormData (namely the results) as a .nc, if that's all you are interested in. This is what the .csv's are currently doing; they only store results, not the covariates/response variables. 

On another note, are you writing unit tests? Because in that case, it would be great if you could add them to the testing suite.

**Comment by eminSerin** [comment]
Thanks a lot for the clarification, Stijn! 
As you said, after loading, the loaded attributes of the train-test set and the actual ones differ (i.e., True with `xarray.Dataset.equals`, but False with `xarray.Dataset.identical`). More importantly, this implies that there's no inherent way to load the train and test sets afterwards without saving the attributes separately and setting them after loading the train-test split. This limitation arises because, regardless of the loading method (e.g., `from_dataframe`, `from_xarray`), the `NormClass` initiation runs `register_batch_effects`, which recalculates the batch effect attributes, resulting in values different from the original train-test split. Given that this is not mentioned anywhere in the manual or technical documentation, I think it might be a critical issue. This is particularly concerning since the `NormClass` manual suggests separating the train-test split outside using scikit-learn and creating `NormData` separately (https://pcntoolkit.readthedocs.io/en/latest/pages/tutorials/01_loading_data.html#pre-processing-and-split-datasets). I can modify "register_batch_effects" so that if these attributes are already provided, it skips recalculating them and submits another PR, along with a better way to save .nc files (without removing any attributes).

**Comment by AuguB** [comment]
Yes, I agree that adding a `NormData.save` method using .nc would be a good addition. It would be great if you could modify the PR to include that.

Regarding the issue with the train-test split outside PCNtoolkit, the manual that you linked to mentions the `make_compatible` method, which is designed specifically to address this issue.
